text/0000-tangle/0000-tangle.md

```diff
@@ -0,0 +1,409 @@
+                tangle: self,
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Insert a vertex into the Tangle, automatically triggering the solidification algorithm.
+    pub fn insert(&mut self, vert: Vertex) -> VertexRefMut {
+        let new_hash = vert.hash;
+        let new_approvees = vert.approvee_hashes();
+
+        // Don't re-insert a vertex
+        if !self.contains(new_hash) {
+            // Perform the tangle insertion
+            self.vertices.insert(new_hash, vert);
+            new_approvees
+                .iter()
+                .for_each(|a| self.txs_to_approvers
+                    .entry(*a)
+                    .or_default()
+                    .push(new_hash));
+
+            // Does the new vertex approve vertices that we don't yet know about?
+            if new_approvees
+                // Do any of the new vertex's approvees...
+                .iter()
+                // ...not exist yet?
+                .any(|approvee| !self.contains(*approvee))
+            {
+                let new_rc = Rc::new(new_hash);
+                // For each approvee of the inserted vertex...
+                let vertices = &self.vertices;
+                let missing_to_approvers = &mut self.missing_to_approvers;
+                new_approvees
+                    .iter()
+                    // ...check to see whether it's missing from the tangle...
+                    .filter(|approvee| !vertices.contains_key(*approvee))
+                        // ...and remember that visiting it is work we need to do later...
+                    .for_each(|approvee| missing_to_approvers
+                        .entry(*approvee)
+                        .or_default()
+                        // ...by associating it with the missing approvee.
+                        .push(new_rc.clone()));
+            }
+
+            // Attempt to propagate solidification information based on the new
+            // information the inserted vertex has provided us with. We do this
+            // by checking to see whether any approvers were waiting upon this vertex.
+            self.missing_to_approvers
+                .remove(&new_hash)
+                .into_iter()
+                .flatten()
+                .filter_map(|hash| Rc::try_unwrap(hash).ok())
+                .for_each(|hash| self.try_solidify(hash));
+        }
+
+        self.get_mut(new_hash)
+            .unwrap() // Can't fail, we just inserted it
+    }
+
+    /// Attempt to perform solidification upon a node (and its approvers). This method is private
+    /// because it is automatically run whenever the tangle is updated with new information
+    fn try_solidify(&mut self, root: TxHash) {
+        // Prevent borrow errors by borrowing the fields independently
+        let vertices = &mut self.vertices;
+        let txs_to_approvers = &self.txs_to_approvers;
+
+        // The algorithm is recursive, but we don't want to use the stack
```

Comment:
this comment is confusing since we are using a stack :)

Reply:
I should definitely clarify what I mean by that comment.